File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/custom/annotation/CorrelationSpec.java
package io.micronaut.mqtt.docs.custom.annotation;

import io.micronaut.mqtt.AbstractMQTTTest;
import io.micronaut.context.ApplicationContext;
import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class CorrelationSpec extends AbstractMQTTTest {

    @Test
    void testUsingACustomAnnotationBinder() {
        ApplicationContext applicationContext = startContext();

// tag::producer[]
        ProductClient productClient = applicationContext.getBean(ProductClient.class);
        productClient.send("a".getBytes());
        productClient.send("b".getBytes());
        productClient.send("c".getBytes());
// end::producer[]

        ProductListener productListener = applicationContext.getBean(ProductListener.class);

        try {
            Set<String> expectedMessages = new HashSet<>();
            expectedMessages.add("a");
            expectedMessages.add("b");
            expectedMessages.add("c");
            await().atMost(5, SECONDS).until(() ->
                    productListener.messages.size() == 3 &&
                    productListener.messages.equals(expectedMessages)
            );
        } finally {
            applicationContext.close();
        }
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/custom/annotation/CorrelationSpec.kt
package io.micronaut.mqtt.docs.custom.annotation

import io.micronaut.mqtt.AbstractMQTTTest
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class CorrelationSpec : AbstractMQTTTest() {

    @Test
    fun testCustomAnnotationBinder() {
        val applicationContext = startContext()

        // tag::producer[]
        val productClient = applicationContext.getBean(ProductClient::class.java)
        productClient.send("a".toByteArray())
        productClient.send("b".toByteArray())
        productClient.send("c".toByteArray())
        // end::producer[]

        val productListener = applicationContext.getBean(ProductListener::class.java)

        try {
            await().atMost(5, TimeUnit.SECONDS).until {
                productListener.messages.size == 3
            }
        } finally {
            applicationContext.close()
        }
    }
}
